export: scaling improvement

esq4 opened the pull request:
In addition to specifying the exact max.size for export, adds an alternative scaling factor.

Parameters are mutually exclusive. When one changes, the other is cleared and not counted in the export.


In addition, each parameter can be reset by clicking the MMB.

![2020-08-24](https://user-images.githubusercontent.com/15331117/90987496-9af7e500-e5b5-11ea-826d-e40508b92301.png)

TurboGit commented:
I'm not sure about this, let's see if there is supporters for adding this. It seems like cluttering the export for little. What I don't like is that the export size and scaling are exclusive and yet present both on the GUI and clearing each other... Maybe just me.

esq4 commented:
> What I don't like is that the export size and scaling are exclusive and yet present both on the GUI and clearing each other

I'll think about it.

esq4 commented:
If such a variant UI?

![image](https://user-images.githubusercontent.com/15331117/92331158-c17e4b80-f09e-11ea-85d9-c1875182ef97.png)

![image](https://user-images.githubusercontent.com/15331117/92331177-e672be80-f09e-11ea-80ab-0b801bfee73a.png)

![image](https://user-images.githubusercontent.com/15331117/92331105-42891300-f09e-11ea-9d0e-c9c5f56c4640.png)

The last settings are saving.

johnny-bit commented:
IMO this is good idea UI-wise now. Seems OK.

For additional points you could throw in dt_calculator_solve in there :) (for both scalling and width/height) or that could be separate PR :)

esq4 commented:
> For additional points you could throw in dt_calculator_solve in there

I was thinking about using a calculator. But this seems redundant to me. And besides that, I tried to make "foolproof".

TurboGit commented on src/common/imageio_module.c:
```diff
@@ -437,6 +437,89 @@ void dt_imageio_insert_storage(dt_imageio_module_storage_t *storage)
+  gchar *scale_str = dt_conf_get_string("plugins/lighttable/export/resizing_factor");
+
+  char sep[4] = "";
+  snprintf( sep, 4, "%g", (double) 3/2);
```
not space after (

TurboGit commented on src/common/imageio_module.c:
```diff
@@ -437,6 +437,89 @@ void dt_imageio_insert_storage(dt_imageio_module_storage_t *storage)
+  dt_conf_set_string("plugins/lighttable/export/resizing_factor", scale_buf);
+
+  free(scale_str);
+
```
superfluous empty line.

TurboGit commented on src/common/imageio_module.c:
```diff
@@ -437,6 +437,89 @@ void dt_imageio_insert_storage(dt_imageio_module_storage_t *storage)
+}
+
+
+void dt_imageio_scale_optim()
```
can  you add some comments, not sure I understand all this business:

esq4 replied:
This is where the string is cleared for better readability: from "1./2.0" is made "1/2".

TurboGit commented on src/libs/export.c:
```diff
@@ -192,11 +202,15 @@ static void export_button_clicked(GtkWidget *widget, dt_lib_export_t *d)
 
   g_free(icc_filename);
+
+  dt_imageio_scale_optim();
```
Can't we have the cale optim routine internal to export.c ? I don't like having that exported as routine of imageio.

esq4 replied:
Yes, it has become possible. Did.

TurboGit commented on src/libs/export.c:
```diff
@@ -213,6 +227,135 @@ void _set_dimensions(dt_lib_export_t *d, int max_width, int max_height)
+  }
+}
+
+static void widht_mdlclick(GtkEntry *spin, GdkEventButton *event, gpointer user_data)
```
widht -> width

Also current dt style is to have all static routines to start with an underscore. I know there is some missing....

TurboGit commented on src/libs/export.c:
```diff
@@ -213,6 +227,135 @@ void _set_dimensions(dt_lib_export_t *d, int max_width, int max_height)
+    height_changed(GTK_EDITABLE(spin), user_data);
+  }
+}
+static void doom_changed(GtkWidget *spin, gpointer user_data)
```
doom???

esq4 replied:
I'm sorry, it was a splinter of mental anguish :-(

TurboGit reviewed (changes requested):
Ok, lot of changes for a first review.

TurboGit commented on src/common/imageio_module.h:
```diff
@@ -231,6 +231,8 @@ int dt_imageio_get_index_of_storage(dt_imageio_module_storage_t *storage);
 /* add a module into the known module list */
 void dt_imageio_insert_storage(dt_imageio_module_storage_t *storage);
 
+void dt_imageio_scale_splitup(double *x, double *y);
```
either this should have the scale string as parameter (and not use the conf string) and/or a comment should be added to clarify the semantic as as-is it is really not clear what this is about.

esq4 replied:
`The function from the string stored in the "plugins/lighttable/export/resizing_factor" parameter of the configuration file generates the x and y fraction's elements to calculate the scaling factor and improve the readability of the string itself displayed in the "scale" field of the export settings.`

Does this comment clarify anything?

TurboGit replied:
Yes, I still don't like this global routine using an hidden prefs. I would prefer having a more generic routine taking a string as parameter and returnin x & y.

TurboGit replied:
You did not change this? Can you explain why? I mean I'd like to understand if there is good arguments to keep it the way it is.

TurboGit commented:
BTW, please rebase on top of current master and do not merge master into your topic branch. TIA.

TurboGit commented on src/libs/export.c:
```diff
@@ -122,6 +132,81 @@ static void _mouse_over_image_callback(gpointer instance, dt_lib_module_t *self)
+static void _scale_optim()
+{
+  double a=1, b=1;
+  dt_imageio_scale_splitup(&a, &b);
```
So my proposal is more something like:

```
gchar *scale_str = dt_conf_get_string(CONFIG_PREFIX "resizing_factor");
dt_imageio_scale_splitup(scale_str, &a, &b);
```

And a, b could probably be renamed num, denum (for numerator denominator).

TurboGit commented on src/common/imageio_module.h:
```diff
@@ -231,6 +231,12 @@ int dt_imageio_get_index_of_storage(dt_imageio_module_storage_t *storage);
 /* add a module into the known module list */
 void dt_imageio_insert_storage(dt_imageio_module_storage_t *storage);
 
+// This function generates the a and b fraction's elements to calculate the scaling factor
```
you probably also to fix a & b in comment.

esq4 replied:
Yes, of course I will fix it. I'm sorry.

TurboGit commented:
Apart from my latest comments I tested it and it is working fine on my side. Will be a nice addition.

TurboGit commented on src/libs/export.c:
```diff
@@ -213,6 +227,135 @@ void _set_dimensions(dt_lib_export_t *d, int max_width, int max_height)
+  const gchar *value = "";
+  value =  gtk_entry_get_text(spin);
+
+  int len = sizeof(value);
```
This is marked as resolved, but it has not been change to const, I suppose this is possible?

TurboGit commented on src/common/imageio_module.h:
```diff
@@ -231,6 +231,12 @@ int dt_imageio_get_index_of_storage(dt_imageio_module_storage_t *storage);
+// and improve the readability of the string itself displayed in the "scale" field of the
+// export settings from the string stored in the "plugins/lighttable/export/resizing_factor"
+// parameter of the configuration file.
+gchar *dt_imageio_resizing_factor_splitup(double *num, double *denum);
```
Again, I asked to have the scale string as parameter here. I still think this is not ok, using an hidden pref as parameter for a public API sounds wrong.

esq4 replied:
This function is quite specific and works with a single configuration parameter.
The current implementation made it possible to exclude an extra line of accessing the configuration from the code.
Your option returns it. But of course I can implement it.

TurboGit reviewed (changes requested):
Last but not least, on #6336 there is a proposal to add a way to select physical units like (inch or centimeter) and a DPI to get a final size in pixels. I suppose this PR will need to be reworked a bit to fit on this new module layout. 

We have a combo names unit with entries : pixel, cm, in

I propose that we add a new entry name : factor

In this mode the label "max size" would be renamed "scaling" with the single entry as done here.

So maybe, just wait a bit for the other PR to be merged as I have asked for some modifications.

esq4 commented:
> So maybe, just wait a bit for the other PR to be merged as I have asked for some modifications.

Well I will wait.

TurboGit commented:
@esq4 : the print-export PR is merged so you can rework this PR on top of master. I would add a new entry in the combobox for support of scaling (with have pixel, cm & inch). And note that this must be only a GUI thing and we want only to have the width & height to be set properly. TIA.

esq4 commented:
OK, wil do.

TurboGit commented:
Perfect thanks! So I'll wait before merging  #6479 to also document the new scaling support.

esq4 commented:
I offer to do so

![image](https://user-images.githubusercontent.com/15331117/95753208-783ba000-0ccb-11eb-8b07-31355f2f3325.png)

![image](https://user-images.githubusercontent.com/15331117/95753291-9d301300-0ccb-11eb-92a7-ae2ea83db405.png)

![image](https://user-images.githubusercontent.com/15331117/95753377-bc2ea500-0ccb-11eb-8c35-beb84807aaee.png)

![image](https://user-images.githubusercontent.com/15331117/95753633-16c80100-0ccc-11eb-8697-342835e3bc77.png)

![image](https://user-images.githubusercontent.com/15331117/95753790-4e36ad80-0ccc-11eb-9411-625c833c5c62.png)

TurboGit commented:
I don't understand why the scale is not one of the options in the combo where we already have cm, inch, pixel? Having another combo just makes this gui more complex than necessary.

esq4 commented:
> I don't understand why the scale is not one of the options in the combo where we already have cm, inch, pixel? Having another combo just makes this gui more complex than necessary.

"Scale" is a multiplier. It's logical to separate it from units of measurement. In addition, when switching the combo "unit" by scrolling, the sizes are recalculated between in, cm and px. And it seemed convenient to me. Switching to scale (unintentional) in this sequence is confusing.

TurboGit commented:
I don't see this this was.

The module is an interface to compute the final size in pixels and only the width in pixels and height in pixels are actually needed by this module. Now we have many ways to specify those values:

- enter directly in pixel
- enter in CM and in this case the size in pixels are displayed but not editable
- enter in INCH and in this case the size in pixels are displayed but not editable

I'm proposing to add a new option:

- enter a scale from the size of the picture and in this case the size in pixels should be displayed but not editable.

This way the UI stay not too saturated of controls.

esq4 commented:
How do you suggest calculating the size in pixels when several images of different sizes are selected? I do not know how.

TurboGit commented:
> And what is the point of displaying the size in pixels when the file is being prepared for printing?

I don't get the question. Why not ? It gives an indication of the actual number of pixels for the exported image as this is the export module (and not the print module) and at the end you have a picture on disk with a number of pixels in width and height.

TurboGit commented:
> How do you suggest calculating the size in pixels when several images of different sizes are selected? I do not know how.

Indeed, that's a good point. We need to take this into account.

TurboGit commented:
> Therefore, it is logically correct to separate these methods in UI.

Maybe in the code, but I still don't agree in the GUI. We may want to have an entry scale in the current combo as I proposed and maybe in this case hide the non editable size in pixels. Then in scale mode we want to record the scale but we probably want to do lot of tweaking afterward as IIRC in imageio code which does the export knows only about max pixels sizes.

esq4 commented:
Maybe so?

![image](https://user-images.githubusercontent.com/15331117/95896366-566b1780-0db6-11eb-8216-2be7530e3ea3.png)

![image](https://user-images.githubusercontent.com/15331117/95896454-700c5f00-0db6-11eb-9b83-a5caf239a7de.png)

![image](https://user-images.githubusercontent.com/15331117/95896526-90d4b480-0db6-11eb-9fd6-0875a4f75994.png)

![image](https://user-images.githubusercontent.com/15331117/95896655-c4174380-0db6-11eb-8ffa-6cae6813f8df.png)

TurboGit commented:
Yes, looks better to me but it looks like you have removed the non editable size in pixels for cm & inch. This is very informative, I don't see why you've removed this. Also even if not used later for other pictures it may make sense to also keep this for scale display. I have really not strong opinion for this last part...

esq4 commented:
I find it superfluous to display a non-editable size in pixels for several reasons:
- when preparing for printing, only the paper size and printer resolution are important. Maximum size in pixels does not specify anything;
- if the user wants to see the maximum size in pixels, then it is enough to select the appropriate item in the combo;
- with all this, it simply takes up space, saturating of the UI of the module.

There is a thought to display this information in a tooltip, but most likely it will be work for the sake of work and is not in demand by anyone.

TurboGit commented:
@esq4 : The display of pixel sizes was done by Aurélien some days ago. I won't remove that except if Aurélien approves this change. @aurelienpierre : Can you cast your vote on this one ? Or anyone else BTW...

esq4 commented:
@TurboGit : Yes, it's better to know his opinion.

aurelienpierre commented:
The actual pixels count is the result of the computation. I just like this kind of double-checking display, both for quick reference and to ensure everything is alright. 

Also, from an user education point of view, it makes very clear that the pixel count is the only thing that actually matters for the export, and DPI × size is merely an interface that produces the pixel count indirectly. Indeed, you will find many users wondering at what DPI they should export for web, or that kind of things that reeks of misunderstanding, so I never let an occasion to use the GUI to teach (because, let's be honest, nobody reads the docs).

esq4 commented:
OK, the approach is clear.

esq4 commented:
![image](https://user-images.githubusercontent.com/15331117/96375811-79763c80-11a5-11eb-81e8-9d9f99237adb.png)

![image](https://user-images.githubusercontent.com/15331117/96375825-90b52a00-11a5-11eb-87d8-9fdbfa7d8655.png)

TurboGit commented:
Some issues:
- The equal pixel displayed in print size (cm or inch) is not refreshed when changing DPI.
- Going from scale to pixels does not refresh the entries (scaling factor is still displayed)
- Select pixel 0 x 0 export OK, then scale 1/2 OK, then select CM and then pixels (just to have the entries selected show 0x0), export and you still have the scale 1/2
- Last but not least, each time I export I have on the console:

```
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[lcms2] error 9: Wrong output color space on transform
unsupported output profile `ISO Coated' has been replaced by sRGB!
[export_job] exported to `/home/obry/Images/Photos/export/tests/img_19_10.jpg'
```

Please be sure to test all possible combinations. Thanks.

TurboGit commented:
Hum, looks like last issue is not due to this PR!

esq4 commented:
> Some issues:
> ...

Thanks, fixed. I have not found others yet.

TurboGit commented:
Still some issues:
- select pixels 1920x1920 export OK
- select scale (let 1) as proposed => export image is 1920 WRONG
- enter 1 in scale (already there but editing seems to reset)
- export OK - size is max
- back to pixels, the size has been reset to 0 x 0 (would be nice to keep last entered 1920x1920).

TurboGit reviewed (changes requested):
Still some issues.

esq4 commented:
> * enter 1 in scale (already there but editing seems to reset)

Yes, it is reset only when editing. When switching, I do not change, since it is very easy to do it through negligence.
I'll think about how to solve this.

esq4 commented:
> > * enter 1 in scale (already there but editing seems to reset)
> 
> Yes, it is reset only when editing. When switching, I do not change, since it is very easy to do it through negligence.
> I'll think about how to solve this.

Removed reseting values when switching. All settins are saving.
Added resetting of dimensions by MMB.

TurboGit commented on src/libs/export.c:
```diff
@@ -902,18 +1135,23 @@ void gui_init(dt_lib_module_t *self)
-  gtk_widget_set_tooltip_text(d->print_width, _("maximum output width\nset to 0 for no scaling"));
+  gtk_widget_set_tooltip_text(d->print_width, _("maximum output width limit.\n"
+                                                "click middle mouse button sets to 0\n"
+                                                " for cancel the constraint"));
```
should be "to cancel..." but I would just drop this line. I would rephrase the line above:

"click middle mouse button to reset to 0"

esq4 replied:
I tried to explain the meaning of this operation. Think don't need it?

TurboGit replied:
Yes I think it is not needed and not proper English.

TurboGit commented on src/libs/export.c:
```diff
@@ -902,18 +1135,23 @@ void gui_init(dt_lib_module_t *self)
   d->print_height = gtk_entry_new();
-  gtk_widget_set_tooltip_text(d->print_height, _("maximum output height\nset to 0 for no scaling"));
+  gtk_widget_set_tooltip_text(d->print_height, _("maximum output height limit.\n"
+                                                 "click middle mouse button sets to 0\n"
```
see above

TurboGit commented on src/libs/export.c:
```diff
@@ -927,45 +1165,70 @@ void gui_init(dt_lib_module_t *self)
-  gtk_widget_set_tooltip_text(d->width, _("maximum output width\nset to 0 for no scaling"));
+  gtk_widget_set_tooltip_text(d->width, _("maximum output width limit.\n"
+                                          "click middle mouse button sets to 0\n"
+                                          " for cancel the constraint"));
```
see above

TurboGit commented on src/libs/export.c:
```diff
@@ -927,45 +1165,70 @@ void gui_init(dt_lib_module_t *self)
   d->height = gtk_entry_new();
-  gtk_widget_set_tooltip_text(d->height, _("maximum output height\nset to 0 for no scaling"));
+  gtk_widget_set_tooltip_text(d->height, _("maximum output height limit.\n"
+                                           "click middle mouse button sets to 0\n"
```
see above

TurboGit commented on src/libs/export.c:
```diff
@@ -927,45 +1165,70 @@ void gui_init(dt_lib_module_t *self)
+  gtk_entry_set_text (GTK_ENTRY(d->scale), dt_conf_get_string(CONFIG_PREFIX "resizing_factor"));
+  gtk_widget_set_tooltip_text(d->scale, _("it can be an integer, decimal fraction or simple fraction.\n"
+                                          "zero or empty values are equal to 1.\n"
+                                          "click the middle mouse button sets the value to 1."));
```
"click the middle mouse button to reset to 1."

TurboGit reviewed (changes requested):
All working fine on my side, just some comments about the English text in tooltip.

TurboGit reviewed (approved):
Looks good to me now. I'll merge now to get more field testing but on my side it is working properly now. Thanks a lot !
